lib/postgres.go

```diff
@@ -0,0 +1,1414 @@
+package lib
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/dgraph-io/badger/v3"
+	"github.com/go-pg/pg/v10"
+	"github.com/go-pg/pg/v10/orm"
+	"github.com/golang/glog"
+)
+
+type Postgres struct {
+	db *pg.DB
+}
+
+func NewPostgres(db *pg.DB) *Postgres {
+	// Print all queries.
+	//db.AddQueryHook(pgdebug.DebugHook{
+	//	Verbose: true,
+	//})
+
+	return &Postgres{
+		db: db,
+	}
+}
+
+func LogSelect(query *orm.Query) error {
+	selectQuery := orm.NewSelectQuery(query)
+	fmter := orm.NewFormatter().WithModel(selectQuery)
+	queryStr, _ := selectQuery.AppendQuery(fmter, nil)
+	glog.Info(string(queryStr))
+	return query.Select()
+}
+
+//
+// Tables
+//
+// When we can, we use unique fields (or combinations of unique fields) as the primary keys on the models.
+// This lets us use the WherePK() query while also minimizing columns and indicies on disk.
+//
+
+type PGChain struct {
+	tableName struct{} `pg:"pg_chains"`
+
+	Name    string `pg:",pk"`
+	TipHash *BlockHash
```

Comment:
Why are some fields not annotated? Explain in your comment at the top.

Reply:
Ok I think I get it. You only annotate when you want to add things to the column definition like pk or use_zero. Still good to maybe explain it at the top. It may be redundant, but if it helps the lazy then it's worth it.

Reply:
So I assume the column names are interpolated into Postgres via magic...? I don't love it, but it's fine for now...

Reply:
Hate magic...

Reply:
Also in your comment, explain in full how something in the migration maps to something in this file, and how it all fits together. Make it as easy as possible for a dev to understand how to fully trace everything.